Add a page on changing the C API (PEP-652) (#682)
python/devguide: c-api.rst

=======================
Changing Python's C API
=======================

The C API is divided into three sections:

1. The internal, private API, available with ``Py_BUILD_CORE`` defined.
   Ideally declared in ``Include/internal/``. Any API named with a leading
   underscore is also considered private.

> 2. The public C API, available when ``Python.h`` is included normally.

Reviewer: Maybe drop _"normally"_?

Author: By *normally* I mean *without defining macros to select the other variants*. I'll expand that in the section below.

exposed only for specific use cases like debuggers and profilers.

With the extern keyword
-----------------------

Functions in ``Include/internal/`` defined with the ``extern`` keyword
*must not and can not* be used outside the CPython code base.  Only
built-in stdlib extensions (built with the ``Py_BUILD_CORE_BUILTIN``
macro defined) can use such functions.

When in doubt, new internal C functions should be defined in
``Include/internal`` using the ``extern`` keyword.

Private names
--------------

Any API named with a leading underscore is also considered internal.
There are two main use cases for using such names rather than putting the
definition in ``Include/internal/`` (or directly in a ``.c`` file):

* Internal helpers for other public API; users should not use these directly;
* “Provisional” API, included in a Python release to test real-world usage
  of new API. Such names should be renamed when stabilized; preferably with
  a macro aliasing the old name to the new one.
  See `"Finalizing the API" in PEP 590`_ for an example.

.. _"Finalizing the API" in PEP 590: https://www.python.org/dev/peps/pep-0590/#finalizing-the-api

.. _public-capi:

Public C API
============

> CPython's public C API if available when ``Python.h`` is included normally.

Reviewer: ```suggestion
CPython's public C API is available when ``Python.h`` is included normally.
```

It should be defined in ``Include/cpython/`` (unless part of the Limited API,
see below).

Guidelines for expanding/changing the public API:

- Make sure the new API follows reference counting conventions. 
  (Following them makes the API easier to reason about, and easier use
  in other Python implementations.)

  - Functions *must not* steal references
  - Functions *must not* return borrowed references
  - Functions returning references *must* return a strong reference

- Make sure the ownership rules and lifetimes of all applicable struct
  fields, arguments and return values are well defined.

Limited API
===========

The Limited API is a subset of the C API designed to guarantee ABI
stability across Python 3 versions.
Defining the macro ``Py_LIMITED_API`` will limit the exposed API to
this subset.

No changes that break the Stable ABI are allowed.

The Limited API should be defined in ``Include/``, excluding the
``cpython`` and ``internal`` subdirectories.

Guidelines for changing the Limited API
---------------------------------------

- Guidelines for the general :ref:`public-capi` apply.

- New Limited API should only be defined if ``Py_LIMITED_API`` is set

>   to a the version the API was added in or higher.

Reviewer: Looks like there's left-overs from an earlier formulation here :)

- All parameter types, return values, struct members, etc. need to be part
  of the Limited API.

  - Functions that deal with ``FILE*`` (or other types with ABI portability
    issues) should not be added.

- Think twice when defining macros.

  - Macros should not expose implementation details
  - Functions must be exported as actual functions, not (only)
    as functions-like macros.
  - If possible, avoid macros. This makes the Limited API more usable in
    languages that don't use the C preprocessor.

- Please start a public discussion before expanding the Limited API

> - The limited API and its intended use must follow standard C, not just

Reviewer: ```suggestion
- The Limited API and its intended use must follow standard C, not just
```

(Nit: I find the _"its intended use"_ part a tiny bit weird. Can it be reformulated?)

Reviewer: Also, the "follow standard C" part is covered in line 130.

    field to your struct, make sure it can be done.
  - Make as few assumptions as possible about implementation details that
    might change in future CPython versions or differ across C API
    implementations. The most important CPython-specific implementation
    details involve:

    - The GIL
    - :ref:`Garbage collection <gc>`
    - Memory layout of PyObject, lists/tuples and other structures

If following these guidelines would hurt performance, add a fast function
(or macro) to the non-limited API and a stable equivalent to the Limited
API.

If anything is unclear, or you have a good reason to break the guidelines,
consider discussing the change at the `capi-sig`_ mailing list.

.. _capi-sig: https://mail.python.org/mailman3/lists/capi-sig.python.org/

Adding a new definition to the Limited API
------------------------------------------

- Add the declaration to a header file directly under ``Include/``, into a
  block guarded with the following:

  .. code-block:: c

    #if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 >= 0x03yy0000

  with the ``yy`` corresponding to the target CPython version, e.g.
  ``0x030A0000`` for Python 3.10.

> - Add an entry to the end of the Stable ABI manifest, ``Misc/stable_abi.txt``.

Reviewer: ```suggestion
- Append an entry to the Stable ABI manifest, ``Misc/stable_abi.txt``.
```
